### PR TITLE
BBR-related fixes

### PR DESCRIPTION
**Pull request by il-theCrow**
- `m_tcb` is no longer a `QuicSocketTxBuffer` attribute, it is now passed as arguments in methods requiring it (`RateSample` related)
- fixed order issues in retransmissions: lost stream frames are added after other text frames (FIFO Tx Scheduler only)
- fixed a problem for which retx frames were all sent in a single packet, exceeding the max allowed size and not following CC/pacing
- fixed a problem for which socket might not have enough data to transmit, causing app-limited `RateSample`s even with applications such as BulkSend

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,12 +1,15 @@
-QUIC implementation for ns-3
+BBR implementation for QUIC in ns-3
 ================================
 
 ## QUIC code base
-This repository contains in the code for a native IETF QUIC implementation in ns-3.
+This repository is a fork of [signetlabdei/quic](https://github.com/signetlabdei/quic) which contains in the code for a native IETF QUIC implementation in ns-3.
```
Comment (mychele) [inline README.md:5]: Rephrase

File: README.md [hunk 2]
```diff
@@ -1,12 +1,15 @@
-QUIC implementation for ns-3
+BBR implementation for QUIC in ns-3
```
Comment (mychele) [inline README.md:1]: QUIC implementation for ns-3

File: README.md [hunk 3]
```diff
@@ -1,12 +1,15 @@
...
-This repository contains in the code for a native IETF QUIC implementation in ns-3.
+This repository is a fork of [signetlabdei/quic](https://github.com/signetlabdei/quic) which contains in the code for a native IETF QUIC implementation in ns-3.
 
 The implementation is described in [this paper](https://arxiv.org/abs/1902.06121).
 
-Please use this [issue tracker](https://github.com/signetlabdei/quic-ns-3/issues) for bugs/questions.
+## BBR implementation
+The implementation provided in this repository is an adaptation of the TCP one provided in [Vivek-anand-jain/ns-3-dev-git/bbr-dev](https://github.com/Vivek-anand-jain/ns-3-dev-git/tree/bbr-dev).
+
+Please use this [issue tracker](https://github.com/il-theCrow/quic-ns-3/issues) for bugs/questions.
```
Comment (mychele) [inline README.md:12]: Move before ## BBR implementation

File: README.md [hunk 4]
```diff
@@ -31,7 +34,7 @@ cd ns-3-dev/src
 
 Then, clone the quic module:
 ```bash
-git clone https://github.com/signetlabdei/quic quic
+git clone https://github.com/il-theCrow/quic quic
```
Comment (mychele) [inline README.md:37]: check

File: model/quic-l4-protocol.cc [hunk 5]
```diff
@@ -680,6 +680,8 @@ QuicL4Protocol::CreateSocket (TypeId congestionTypeId)
   socket->SetNode (m_node);
   socket->SetQuicL4 (this);
 
+  socket->InitializeScheduling();
```
Comment (mychele) [inline model/quic-l4-protocol.cc:683]: It seems that this PR also merges the scheduler code. Are we ok with this? @fedech

File: model/quic-socket-base.cc [hunk 6]
```diff
@@ -221,12 +222,22 @@ QuicSocketBase::GetTypeId (void)
...
+				   "Default latency bound for the EDF scheduler",
+                   TimeValue (MilliSeconds (100)),
+                   MakeTimeAccessor (&QuicSocketBase::m_defaultLatency),
+                   MakeTimeChecker ())
+    .AddAttribute (
+                   "LegacyCongestionControl",
+                   "When true, use TCP implementations for the congestion control",
+                   BooleanValue (false),
+                   MakeBooleanAccessor (&QuicSocketBase::m_quicCongestionControlLegacy),
+                   MakeBooleanChecker ())
```
Comment (mychele) [inline model/quic-socket-base.cc]: Check indent

File: model/quic-socket-base.cc [hunk 7]
```diff
@@ -221,12 +222,22 @@ QuicSocketBase::GetTypeId (void)
...
+				   TypeIdValue (QuicSocketTxFifoScheduler::GetTypeId ()),
+				   MakeTypeIdAccessor (&QuicSocketBase::m_schedulingTypeId),
+				   MakeTypeIdChecker ())
+	.AddAttribute ("DefaultLatency",
+				   "Default latency bound for the EDF scheduler",
+                   TimeValue (MilliSeconds (100)),
+                   MakeTimeAccessor (&QuicSocketBase::m_defaultLatency),
+                   MakeTimeChecker ())
+    .AddAttribute (
+                   "LegacyCongestionControl",
```
Comment (mychele) [inline model/quic-socket-base.cc]: same line?

File: model/quic-socket-base.cc [hunk 8]
```diff
@@ -296,6 +307,16 @@ QuicSocketBase::GetTypeId (void)
                      "Receive QUIC packet from UDP protocol",
                      MakeTraceSourceAccessor (&QuicSocketBase::m_rxTrace),
                      "ns3::QuicSocketBase::QuicTxRxTracedCallback")
+	.AddAttribute ("CongControl",
```
Comment (mychele) [inline model/quic-socket-base.cc]: Do not mix attributes and sources, place all the attributes before the sources

File: model/quic-socket-base.cc [hunk 9]
```diff
@@ -296,6 +307,16 @@ QuicSocketBase::GetTypeId (void)
...
+				   "The Object managing CC operations",
+				   PointerValue (),
```
Comment (mychele) [inline model/quic-socket-base.cc]: No default?

File: model/quic-socket-base.cc [hunk 10]
```diff
@@ -296,6 +307,16 @@ QuicSocketBase::GetTypeId (void)
...
+				   MakePointerAccessor (&QuicSocketBase::m_congestionControl),
+				   MakePointerChecker<QuicCongestionOps> ())
+	.AddAttribute ("TCB",
+				   "The connection's QuicSocketState",
+				   PointerValue (),
```
Comment (mychele) [inline model/quic-socket-base.cc]: No default?

File: model/quic-socket-base.cc [hunk 11]
```diff
@@ -536,6 +557,8 @@ QuicSocketBase::QuicSocketBase (void)
 
   ok = m_tcb->TraceConnectWithoutContext ("HighestSequence",
                                           MakeCallback (&QuicSocketBase::UpdateHighTxMark, this));
+  NS_ASSERT (ok == true);
```
Comment (mychele) [inline model/quic-socket-base.cc:555]: Maybe use an NS_ASSERT_MSG(ok == true, "message to explain the error");

File: model/quic-socket-base.cc [hunk 12]
```diff
@@ -3071,6 +3109,22 @@ QuicSocketBase::GetInitialPacketSize () const
   return m_initialPacketSize;
 }
 
+void QuicSocketBase::SetLatency(uint32_t streamId, Time latency) {
```
Comment (mychele) [inline model/quic-socket-base.cc:3107]: maybe run a check-style.py, these are not consistent with the ns-3 code style

File: model/quic-socket-base.h [hunk 13]
```diff
@@ -551,6 +559,37 @@ class QuicSocketBase : public QuicSocket
   virtual enum SocketErrno GetErrno (void) const;
   virtual enum SocketType GetSocketType (void) const;
 
+	/**
```
Comment (mychele) [inline model/quic-socket-base.h:563]: Check indent (probably solved by check-style.py)

File: model/quic-socket-tx-buffer.cc [hunk 14]
```diff
@@ -34,796 +34,589 @@
...
-NS_OBJECT_ENSURE_REGISTERED (QuicSocketTxBuffer);
-
-TypeId
-QuicSocketTxBuffer::GetTypeId (void)
-{
-  static TypeId tid =
-    TypeId ("ns3::QuicSocketTxBuffer").SetParent<Object> ().SetGroupName (
-      "Internet").AddConstructor<QuicSocketTxBuffer> ()
+TypeId QuicSocketTxItem::GetTypeId(void) {
+	static TypeId tid =
```
Comment (mychele) [inline model/quic-socket-tx-buffer.cc:46]: fix indent and formatting

File: model/quic-socket-tx-buffer.cc [hunk 15]
```diff
@@ -34,796 +34,589 @@
...
+}
+
+QuicSocketTxItem::QuicSocketTxItem(const QuicSocketTxItem &other) :
+		m_packet(other.m_packet), m_packetNumber(other.m_packetNumber), m_lost(
+				other.m_lost), m_retrans(other.m_retrans), m_sacked(
+				other.m_sacked), m_acked(other.m_acked), m_isStream(
+				other.m_isStream), m_isStream0(other.m_isStream0), m_lastSent(
+				other.m_lastSent), m_generated(other.m_generated) {
+
+}
```
Comment (mychele) [inline model/quic-socket-tx-buffer.cc:71]: fix indent and formatting

File: model/quic-socket-tx-buffer.cc [hunk 16]
```diff
@@ -34,796 +34,589 @@
 #include "ns3/abort.h"
```
Comment (mychele) [inline model/quic-socket-tx-buffer.cc:35]: This file is very hard to review, formatting issues make it look completely different from the old one. Please check!

File: model/quic-socket-tx-buffer.h [hunk 17]
```diff
@@ -37,282 +37,322 @@
 namespace ns3 {
```
Comment (mychele) [inline model/quic-socket-tx-buffer.h:38]: Fix indent throughout

File: model/quic-socket-tx-buffer.h [hunk 18]
```diff
@@ -37,282 +37,322 @@
...
-  uint32_t      m_priorAckBytesSent {0}; //!< amount of ACK-only bytes sent up to a flight ago
-  uint8_t       m_ackBytesSentWin {0};
+class QuicSocketTxScheduler;
+
+struct RateSample {
+	DataRate m_deliveryRate;   //!< The delivery rate sample
+	bool m_isAppLimited { false }; //!< Indicates whether the rate sample is application-limited
+	Time m_interval;       //!< The length of the sampling interval
+	uint32_t m_delivered { 0 }; //!< The amount of data marked as delivered over the sampling interval
+	uint32_t m_priorDelivered { 0 }; //!< The delivered count of the most recent packet delivered
```
Comment (mychele) [inline model/quic-socket-tx-buffer.h]: {0};

File: model/quic-socket-tx-edf-scheduler.cc [hunk 19]
```diff
@@ -0,0 +1,328 @@
...
+#include "quic-socket-base.h"
+
+namespace ns3 {
+
+NS_LOG_COMPONENT_DEFINE("QuicSocketTxEdfScheduler");
+
+NS_OBJECT_ENSURE_REGISTERED(QuicSocketTxEdfScheduler);
+
+TypeId QuicSocketTxEdfScheduler::GetTypeId(void) {
+	static TypeId tid = TypeId("ns3::QuicSocketTxEdfScheduler").SetParent<
```
Comment (mychele) [inline model/quic-socket-tx-edf-scheduler.cc:45]: fix formatting

File: model/quic-socket-tx-edf-scheduler.cc [hunk 20]
```diff
@@ -0,0 +1,328 @@
+/* -*- Mode:C++; c-file-style:"gnu"; indent-tabs-mode:nil; -*- */
```
Comment (mychele) [inline model/quic-socket-tx-edf-scheduler.cc:1]: Indent and formatting throughout (check-style will probably be enough)

File: model/quic-socket-tx-edf-scheduler.cc [hunk 21]
```diff
@@ -0,0 +1,328 @@
...
+	//Print(std::cout);
+
+	return outItem;
+}
+
+void QuicSocketTxEdfScheduler::SetLatency(uint32_t streamId, Time latency) {
+	m_latencyMap[streamId] = latency;
+}
+
+Time QuicSocketTxEdfScheduler::GetLatency(uint32_t streamId) {
```
Comment (mychele) [inline model/quic-socket-tx-edf-scheduler.cc:298]: Get methods should be const

File: model/quic-socket-tx-edf-scheduler.cc [hunk 22]
```diff
@@ -0,0 +1,328 @@
...
+	Time latency = m_defaultLatency;
+	try {
```
Comment (mychele) [inline model/quic-socket-tx-edf-scheduler.cc:300]: do not use try/catch, use ns-3 asserts or aborts

File: model/quic-socket-tx-edf-scheduler.h [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+	 */
+	void SetDefaultLatency(Time latency);
+
+	/**
+	 * Get the default latency bound
+	 *
+	 * \param streamId The stream ID
+	 * \return The default maximum latency
+	 */
+	Time GetDefaultLatency();
```
Comment (mychele) [inline model/quic-socket-tx-edf-scheduler.h:94]: const

File: model/quic-socket-tx-pfifo-scheduler.cc [hunk 24]
```diff
@@ -0,0 +1,301 @@
+/* -*- Mode:C++; c-file-style:"gnu"; indent-tabs-mode:nil; -*- */
+/*
+ * Copyright (c) 2019 SIGNET Lab, Department of Information Engineering, University of Padova
```
Comment (mychele) [inline model/quic-socket-tx-pfifo-scheduler.cc:3]: Update

File: model/quic-socket-tx-pfifo-scheduler.cc [hunk 25]
```diff
@@ -0,0 +1,301 @@
+/* -*- Mode:C++; c-file-style:"gnu"; indent-tabs-mode:nil; -*- */
```
Comment (mychele) [inline model/quic-socket-tx-pfifo-scheduler.cc:1]: Fix formatting and indent

File: model/quic-socket-tx-pfifo-scheduler.cc [hunk 26]
```diff
@@ -0,0 +1,301 @@
...
+				other.m_retxFirst) {
+	m_appList = other.m_appList;
+}
+
+QuicSocketTxPFifoScheduler::~QuicSocketTxPFifoScheduler(void) {
+	m_appList = QuicTxPacketList();
+	m_appSize = 0;
+}
+
+void QuicSocketTxPFifoScheduler::Print(std::ostream &os) const {
```
Comment (mychele) [inline model/quic-socket-tx-pfifo-scheduler.cc:72]: Can this method be implemented in the base class?

File: model/quic-socket-tx-scheduler.cc [hunk 27]
```diff
@@ -0,0 +1,288 @@
+/* -*- Mode:C++; c-file-style:"gnu"; indent-tabs-mode:nil; -*- */
```
Comment (mychele) [inline model/quic-socket-tx-scheduler.cc:1]: format, indent, copyright

File: model/quic-socket-tx-scheduler.h [hunk 28]
```diff
@@ -0,0 +1,132 @@
...
+public:
+	/**
+	 * \brief Get the type ID.
+	 * \return the object TypeId
+	 */
+	static TypeId GetTypeId(void);
+
+	QuicSocketTxScheduler();
+	QuicSocketTxScheduler(const QuicSocketTxScheduler &other);
+	~QuicSocketTxScheduler(void); /**
```
Comment (mychele) [inline model/quic-socket-tx-scheduler.h:51]: virtual

Comment (mychele) [reply model/quic-socket-tx-scheduler.h:51]: and line break for the doxy of the other method

File: model/quic-socket-tx-scheduler.h [hunk 29]
```diff
@@ -0,0 +1,132 @@
...
+	 * Print the scheduler buffer information to a string
+	 *
+	 * \param os the std::ostream object
+	 */
+	virtual void Print(std::ostream &os) const
```
Comment (mychele) [inline model/quic-socket-tx-scheduler.h:56]: You can make it a pure abstract function
`virtual void Print(std::ostream &os) const = 0;`

File: model/quic-socket-tx-scheduler.h [hunk 30]
```diff
@@ -0,0 +1,132 @@
...
+	}
+
+	//friend std::ostream & operator<< (std::ostream & os, QuicSocketTxBuffer const & quicTxBuf);
+
+	/**
+	 * Add a tx item to the scheduling list
+	 *
+	 * \param item a smart pointer to a transmission item
+	 */
+	virtual void Add(Ptr<QuicSocketTxItem> item, bool retx)
```
Comment (mychele) [inline model/quic-socket-tx-scheduler.h:68]: same here = 0;

File: model/quic-socket-tx-scheduler.h [hunk 31]
```diff
@@ -0,0 +1,132 @@
...
+		NS_UNUSED(retx);
+	}
+
+	/**
+	 * \brief Get the next scheduled packet with a specified size
+	 *
+	 * \param numBytes number of bytes of the QuicSocketTxItem requested
+	 * \return the item that contains the right packet
+	 */
+	virtual Ptr<QuicSocketTxItem> GetNewSegment(uint32_t numBytes)
```
Comment (mychele) [inline model/quic-socket-tx-scheduler.h:80]: same

**Review by mychele — CHANGES_REQUESTED** [review]
Please check and fix the formatting of the code. I could not review a couple of files.

Besides, feel free to update the authors for the files you modified.

File: model/quic-socket-base.cc (hunk 6 above)
Comment (il-theCrow) [reply model/quic-socket-base.cc]: fixed in 331467b

File: model/quic-socket-base.cc (hunk 7 above)
Comment (il-theCrow) [reply model/quic-socket-base.cc]: fixed in 331467b

File: model/quic-socket-base.cc (hunk 8 above)
Comment (il-theCrow) [reply model/quic-socket-base.cc]: fixed in 331467b

File: model/quic-socket-base.cc (hunk 10 above)
Comment (il-theCrow) [reply model/quic-socket-base.cc]: attribute removed in 331467b

File: model/quic-socket-tx-buffer.h (hunk 18 above)
Comment (il-theCrow) [reply model/quic-socket-tx-buffer.h]: fixed in 331467b

**Comment by fedech** [comment]
Superceded by new PR

**Pull request closed without merging**